bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/ADBridgeHandler.java

```diff
@@ -0,0 +1,427 @@
+ * Based partly on and including code from the original OH1 alarmdecoder binding by Bernd Pfrommer.
+ *
+ * @author Bernd Pfrommer - Initial contribution (OH1 version)
+ * @author Bob Adair - Re-factored into OH2 binding
+ */
+@NonNullByDefault
+public abstract class ADBridgeHandler extends BaseBridgeHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(ADBridgeHandler.class);
+
+    protected @Nullable BufferedReader reader = null;
+    protected @Nullable BufferedWriter writer = null;
+    protected @Nullable Thread msgReaderThread = null;
+    private final Object msgReaderThreadLock = new Object();
+    protected @Nullable AlarmDecoderDiscoveryService discoveryService;
+    protected boolean discovery;
+    protected boolean panelReadyReceived = false;
+    protected volatile @Nullable Date lastReceivedTime;
+    protected volatile boolean writeException;
+
+    protected @Nullable ScheduledFuture<?> connectionCheckJob;
+    protected @Nullable ScheduledFuture<?> connectRetryJob;
+
+    public ADBridgeHandler(Bridge bridge) {
+        super(bridge);
+    }
+
+    @Override
+    public void dispose() {
+        logger.trace("dispose called");
+        disconnect();
+    }
+
+    @Override
+    public Collection<Class<? extends ThingHandlerService>> getServices() {
+        return Collections.singletonList(BridgeActions.class);
+    }
+
+    public void setDiscoveryService(AlarmDecoderDiscoveryService discoveryService) {
+        this.discoveryService = discoveryService;
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        // Accepts no commands, so do nothing.
+    }
+
+    /**
+     * Send a command to the alarm decoder using a buffered writer. This could block if the buffer is full, so it should
+     * eventually be replaced with a queuing mechanism and a separate writer thread.
+     *
+     * @param command Command string to send including terminator
+     */
+    public void sendADCommand(ADCommand command) {
+        logger.debug("Sending AD command: {}", command);
+        try {
+            if (writer != null) {
+                writer.write(command.toString());
+                writer.flush();
+            }
+        } catch (IOException e) {
+            logger.info("Exception while sending command: {}", e.getMessage());
+            writeException = true;
+        }
+    }
+
+    protected abstract void connect();
+
+    protected abstract void disconnect();
+
+    protected void scheduleConnectRetry(long waitMinutes) {
+        logger.debug("Scheduling connection retry in {} minutes", waitMinutes);
+        connectRetryJob = scheduler.schedule(this::connect, waitMinutes, TimeUnit.MINUTES);
+    }
+
+    protected void startMsgReader() {
+        synchronized (msgReaderThreadLock) {
+            msgReaderThread = new Thread(this::readerThread, "AD Reader");
```

Comment:
Please make sure to set the thread as daemon.

Reply:
Done.